service/history/queuev2/virtual_queue.go

```diff
@@ -436,6 +419,81 @@ func (q *virtualQueueImpl) loadAndSubmitTasks() {
 	}
 }
 
+func (q *virtualQueueImpl) InsertSingleTask(task task.Task) bool {
+	q.Lock()
+	defer q.Unlock()
+
+	taskKey := task.GetTaskKey()
+	var slice VirtualSlice
+
+	for e := q.virtualSlices.Front(); e != nil; e = e.Next() {
+		s := e.Value.(VirtualSlice)
+		r := s.GetState().Range
+		if taskKey.Compare(r.InclusiveMinTaskKey) >= 0 && taskKey.Compare(r.ExclusiveMaxTaskKey) < 0 {
+			slice = s
+			break
+		}
+	}
+
+	if slice == nil {
+		// the new task is outside of the current range, it will be read from the DB on the next poll
+		return false
+	}
+
+	slice.InsertTask(task)
+	q.monitor.SetSlicePendingTaskCount(slice, slice.GetPendingTaskCount())
+
+	now := q.timeSource.Now()
+	if err := q.submitTask(now, task); err != nil {
+		q.logger.Error("Virtual queue failed to submit task", tag.Error(err))
+		return false
+	}
+
+	return true
+}
+
+func (q *virtualQueueImpl) RemoveScheduledTasksAfter(key persistence.HistoryTaskKey) {
+	q.Lock()
+	defer q.Unlock()
+
+	for e := q.virtualSlices.Front(); e != nil; e = e.Next() {
+		slice := e.Value.(VirtualSlice)
+		slice.CancelTasksAfter(key)
+		q.monitor.SetSlicePendingTaskCount(slice, slice.GetPendingTaskCount())
+	}
+}
+
+func (q *virtualQueueImpl) submitTask(now time.Time, task task.Task) error {
+	if persistence.IsTaskCorrupted(task) {
+		q.logger.Error("Virtual queue encountered a corrupted task", tag.Dynamic("task", task))
+		q.metricsScope.IncCounter(metrics.CorruptedHistoryTaskCounter)
+
+		task.Ack()
+		return nil
+	}
+
+	scheduledTime := task.GetTaskKey().GetScheduledTime()
+	// if the scheduled time is in the future, we need to reschedule the task
+	if now.Before(scheduledTime) {
+		q.rescheduler.RescheduleTask(task, scheduledTime)
+		return nil
+	}
+	// shard level metrics for the duration between a task being written to a queue and being fetched from it
+	q.metricsScope.RecordHistogramDuration(metrics.TaskEnqueueToFetchLatency, now.Sub(task.GetVisibilityTimestamp()))
+	task.SetInitialSubmitTime(now)
+	submitted, err := q.processor.TrySubmit(task)
+	if err != nil {
+		return err
```

Comment:
This is not correct. It should be submitted to rescheduler as long as it's not accepted by the processor regardless the error, because the pending task tracker has tracked the task when it's loaded from the database